go-controller/pkg/ovn/pods.go

```diff
@@ -409,6 +409,33 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) (err error) {
 
 	}
 
+	if annotation == nil {
+		podAnnotation := util.PodAnnotation{
+			IPs: podIfAddrs,
+			MAC: podMac,
+		}
+		var nodeSubnets []*net.IPNet
+		if nodeSubnets = oc.lsManager.GetSwitchSubnets(logicalSwitch); nodeSubnets == nil {
+			return fmt.Errorf("cannot retrieve subnet for assigning gateway routes for pod %s, node: %s",
+				pod.Name, logicalSwitch)
+		}
+		err = oc.addRoutesGatewayIP(pod, &podAnnotation, nodeSubnets)
+		if err != nil {
+			return err
+		}
+		var marshalledAnnotation map[string]string
+		marshalledAnnotation, err = util.MarshalPodAnnotation(&podAnnotation)
+		if err != nil {
+			return fmt.Errorf("error creating pod network annotation: %v", err)
+		}
+
+		klog.V(5).Infof("Annotation values: ip=%v ; mac=%s ; gw=%s\nAnnotation=%s",
+			podIfAddrs, podMac, podAnnotation.Gateways, marshalledAnnotation)
+		if err = oc.kube.SetAnnotationsOnPod(pod, marshalledAnnotation); err != nil {
+			return fmt.Errorf("failed to set annotation on pod %s: %v", pod.Name, err)
```

Comment:
Since we release the allocated IPs on error, we now need to remove the annotation on error as well since we might get different IPs the next time we try, right?

Except, will the node side handle that correctly?

Reply:
Good point. Will look into that.

Reply:
Oh, that's annoying; if we allocate everything correctly in OVN and just fail to set the annotations, it would be better to just retry setting the annotation. We even handle that case (`addLogicalPort` checks to see if it already exists...)

Reply:
@squeed the problem is not trying to set it again. The problem is the pod is usually gone at this point and we have wasted OVN operations, and have to issue more OVN operations to rollback things.

Reply:
@danwinship so explored the path you mentioned. If the annotation is set when we add logical port again, we will try to configure the IP on the port in OVN. However, there is a hole here where we never try to allocate the IP with the allocator, so another pod that comes in after could get the same IP and we could end up with duplicate IPs.

I'm thinking we should rollback the annotation, and I'll also add a commit to cover the aforementioned case by check if we can allocate the IP specified and if not, error out.

Reply:
Alternatively, we could just try re-fetching the pod from k8s at the top of `addLogicalPort`, and if it fails, bail out, and if it succeeds, assume that it's safe to do all the OVN operations, and then if the set-annotations call fails at the end, we just roll everything back like before, but hopefully that will happen less often now since we're doing the check at the start.

Reply:
@danwinship Isn't setting the annotation equivalent of doing a pod-check while doing work? My concern is additional kube API calls. Those have a non-zero penalty. The only time we will see that the pod is already deleted by the time the master gets to annotating it is:

- scaling up and we can't keep up with CNI requests on the node
- we have seen this on e2e tests during namespace deletion.

Are there other cases? 

It's in some ways a win-some lose-some problem. Either we have to roll back annotations or roll back ovn transactions when the pod errors out. We may solve the problem in `addLogicalPort` by trying to annotate early, but there will be cases where the annotation succeeds and so `addLogicalPort` doesn't error out, but it is immediately followed by `deleteLogicalPort` and all that work we will do in doing ovn transactions will be for nothing.

Reply:
"We may solve the problem in addLogicalPort by trying to annotate early, but there will be cases where the annotation succeeds and so addLogicalPort doesn't error out, but it is immediately followed by deleteLogicalPort and all that work we will do in doing ovn transactions will be for nothing."

This is the same behavior as there is today. If we add logical port we have done all the OVN commands, fail to set annotation, then receive delete event and deleteLogicalPort. There's nothing we can really do about this. This patch however will lessen the amount of unnecessary OVN commands we issue (because the pod is already deleted) and stop the persistent errors we see in the logs of "failed to set annotation".

Reply:
> This is the same behavior as there is today.

No, I was saying, add an extra k8s API check at the top. Then if the pod gets deleted after we queue the Add event but before we process the event, then we'll notice before doing any OVN work and can abort immediately. If the pod still exists at the top of `addLogicalPort` then we're _probably_ good. Of course, it's theoretically possible it will be deleted _during_ `addLogicalPort`, so we still need to keep the existing cleanup code, but it just won't get run as often.

Reply:
I decided to go a different route. After writing the code I realized there's no real need to rollback annotations, we just need to make ovn-kubernetes smarter about handling annotations and IP allocations. The new patch never releases IPs, unless the annotation fails, meaning that as long as we allocate IPs and can set annotation, those stay for the lifecycle of the pod. If OVN configuration later fails, we eventually retry on an update, and we just check the existing annotation, and ensure the IPs are already allocated. If something goes wrong, we allocate new IPs and update accordingly.

Reply:
@trozet like @squeed said

```
> Oh, that's annoying; if we allocate everything correctly in OVN and just fail to set the annotations, it would be better to just retry setting the annotation. We even handle that case (`addLogicalPort` checks to see if it already exists...)
```
we should re-try few times if setting annotation on the pod fails. If the pod itself is gone, then the re-try should exit sooner and execute rollback OVN commands. If the rollback OVN commands fails, that is ok since stale entries in OVN is less harmer than not having required OVN entries for the running pods or live K8s resources. 

We any ways have plans for periodic syncing to remove those stale OVN data. Pod Annotation is a nice `barrier` for the worker nodes to know whether they can proceed with the plumbing of the Pod interface or not.

Reply:
I don't think we see failures where the API server has gone down and we cannot annotate. Can we hold off on doing that unless we actually see that is a consistent problem?

Reply:
We saw those during annotating nodes. But that's something we might run into. I agree we can hold off on doing that till we see a consistent problem.

Reply:
Pretty sure we can get transient annotation errors during OCP installs at the point when the bootstrap node goes away (if we were still talking to the bootstrap apiserver instead of one of the new ones).

Reply:
I'll defer handling annotation retries in a subsequent PR.